File: DAY-12/topkfrequent.js
class Pair {
    constructor(key, count) {
        this.key = key;
        this.count = count;
    }
}
function topKFrequent(nums, k) {
    let freq = new Map();
    for(val of nums) {
        freq.set(val, (freq.get(val) || 0) + 1);
    }
    let pairs = [];
    for(const[key, count] of freq) {
        console.log(" key is ", key, " count is ", count);
        pairs.push({key, count});
    }
    pairs.sort((a, b) => b.count - a.count);
    let result = [];
    for(let i=0; i < k; i++) {
        result.push(pairs[i].key);
    }
    return result;
}

console.log(topKFrequent([1, 1, 1, 2, 2, 3], 2));

File: DSA SHEET/DAY-12/topKfrequent.java
import java.util.*;
public class topKfrequent {
    static class Pair {
        int key;
        int count;
        public Pair(int key, int count) {
            this.key = key;
            this.count = count;
        }
    }
    public static List<Integer> topKfrequentNumber(int[] nums, int k) {
        HashMap<Integer, Integer> map = new HashMap<>();
        for(int val : nums) {
            map.put(val, map.getOrDefault(val, 0) + 1);
        }
        PriorityQueue<Pair> pq = new PriorityQueue<>((a, b) -> b.count - a.count);
        for(Integer key : map.keySet()) {
            pq.add(new Pair(key, map.get(key)));
        }
        List<Integer> list = new ArrayList<>();
        for(int i=0; i < k; i++) {
            list.add(pq.poll().key);
        }
        return list;
    }
    public static void main(String[] args) {
        System.out.println(topKfrequentNumber(new int[]{1, 1, 1, 2, 2, 3}, 2));
    }
}
